### PR TITLE
Feat/trn 967/highlighter do not use normalize on root

### PR DESCRIPTION
**Pull request by olga-kulish**
Related to https://oat-sa.atlassian.net/browse/TRN-967
Related to https://oat-sa.atlassian.net/browse/TRN-894

Use it with https://github.com/oat-sa/tao-deliver-fe/pull/110 

See PR comment there about why changes were needed

**How to run unit test** 
[for myself some tests in `npm run test` fail, may be windows issue]:
`npx qunit-testrunner --spec "**/highlighter/test.html"` 
to view all test cases, not just errors, in `.qunit-testrunner.config.json` set `"verbose": 4`
run `npm run build` before running test, if code (not test) was changed

### CODE REVIEW

File: src/highlighter.js
Comment (oatymart) [inline src/highlighter.js:455]: Please add JSDoc for new return value.

File: src/highlighter.js
Comment (oatymart) [inline src/highlighter.js:645]: This might depend on saving & restoring highlights (TRN-894), right? Do you think we need to check that before finalising this PR?

File: src/highlighter.js
Comment (oatymart) [inline src/highlighter.js:316]:
```suggestion
                                    addSplitData(wrapperNode, splitData.beforeWasSplit, splitData.afterWasSplit);
```

File: src/highlighter.js
Comment (oatymart) [inline src/highlighter.js:311]: The condition is pretty hard to read. Is it possible to extract some of the inner parts to named variables?

File: src/highlighter.js
Comment (oatymart) [inline src/highlighter.js:149]: Suggestion: A case for a small helper function? `removeEmptyTextNode(node, nodeToCompare)`

File: src/highlighter.js
Comment (oatymart) [inline src/highlighter.js:116]: I'm not sure about putting that condition _inside_ this function. Since you are checking `if (keepEmptyNodes)` (or its inverse) some 10 times in the module, may as well add it _around_ `addSplitData()` in the 2-3 places where needed. The number of conditions in the code would increase, but it seems easier to understand when they are all on the same level. WDYT?

**Review by oatymart — COMMENTED** [review]
I've tested this a bit on Terre. It seems working like before, and unit tests pass, but I'll still test a bit more thoroughly.

We'll also need to see some unit tests added, to cover the behaviour which is changed using your new parameter. Let me know if you need any help writing/running tests, or running Terre.

File: src/highlighter.js
Comment (olga-kulish) [reply src/highlighter.js:645]: We agreed last week to not merge this before TRN-895 is done.
Sad that it delays testing, but you can see in github actions that pipeline can't build tao-deliver-fe with git-branch dependency

File: src/highlighter.js
Comment (olga-kulish) [reply src/highlighter.js:116]: Agree

File: src/highlighter.js
Comment (olga-kulish) [reply src/highlighter.js]: I pushed changes to the same PR both here and in `tao-deliver-fe` - because in TRN-894 I added a 'model', so implementation of TRN-967 in `tao-deliver-fe` also changed a lot

**Review by oatymart — COMMENTED** [review]
The feature is looking very good now, both on Terre and what's been done so far on Solar.
I can't find any real issues with the code changes here, the new functionality is well-isolated from the old. Well done! There are some tests not passing yet, hopefully it's a minor fix to make.

File: src/highlighter.js
Comment (olga-kulish) [inline src/highlighter.js:168]: FYI: If you do a double-click, one word is selected. If you do a triple-click, whole element is selected. This selection is a bit weird: for
```
  <div>
    <h2>here</h2>  <!-- triple-click here -->
    <p>there</p>
  </div>
```
range will be 
```
commonAncestorContainer: '<div>' element node
startContainer: 'here' text node
startOffset: 0
endContainer: '<p>' element node
endOffset: 0
```

Such selection isn't handled well by existing code - it continues to highlight until the end of `range.commonAncestorContainer`
https://screenrec.com/share/60Fl9QfbhM
https://screenrec.com/share/Ycn24KzAqG

Same happens not only on triple-click, but also for some mouse positions, where you mouseup below your element.

File: src/highlighter.js
Comment (olga-kulish) [reply src/highlighter.js:235]: There is a comment above `addSplitData`:
_'Use it when deleting this highlight to know if highlight content should be merged with neighbour text nodes or not.'_

File: test/highlighter/test.js
Comment (olga-kulish) [reply test/highlighter/test.js:1908]: I kind of wanted to check that after merging/splitting highlights exact value is copied from original highlight to resulting one. With `true/false` I would ideally need more test cases to check that it copies value, not recalculates,

File: test/highlighter/test.js
Comment (olga-kulish) [reply test/highlighter/test.js:1863]: It passes for me now - please do `npm run build` and run test again

**Review by oatymart — APPROVED** [review]
Works well. Didn't find any regressions.

- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Pull request's target is not `master`
- [ ] Documentation is updated according to changes (if applicable)

File: src/highlighter.js
Comment (shaveko) [inline src/highlighter.js:1067]: I'd mention whitelist in jsdoc here

File: src/highlighter.js
Comment (shaveko) [inline src/highlighter.js:785]: Why you've chosen to use `c` intead of full name `color`?

**Review by shaveko — APPROVED** [review]
That's quite hardcore stuff, I can say. 

Reviewed mostly the code style and general approah. Please check my 2 commetns for optional code updates

File: src/highlighter.js
Comment (olga-kulish) [reply src/highlighter.js:785]: This name is used by another branch (`!keepEmptyNodes`), so for consistency - not that this consistency has any meaning now, since format is completely different. But let's keep it so we won't have to retest it :)

**Review by shaveko — APPROVED** [review]
Looks good

**Pull request merged**
